pkg/virtctl/vmexport/vmexport.go

```diff
@@ -857,3 +897,131 @@ func getExportSecretName(vmexportName string) string {
 func errExportAlreadyExists(err error) bool {
 	return strings.Contains(err.Error(), "VirtualMachineExport") && strings.Contains(err.Error(), "already exists")
 }
+
+// Port-forward functions
+
+// translateServicePortToTargetPort tranlates the specified port to be used with the service's pod
+func translateServicePortToTargetPort(localPort string, remotePort string, svc k8sv1.Service, pod k8sv1.Pod) ([]string, error) {
+	ports := []string{}
+	portnum, err := strconv.Atoi(remotePort)
+	if err != nil {
+		return ports, err
+	}
+	containerPort, err := kubectlutil.LookupContainerPortNumberByServicePort(svc, pod, int32(portnum))
+	if err != nil {
+		// can't resolve a named port, or Service did not declare this port, return an error
+		return ports, err
+	}
+
+	// convert the resolved target port back to a string
+	remotePort = strconv.Itoa(int(containerPort))
+	if localPort != remotePort {
+		return append(ports, fmt.Sprintf("%s:%s", localPort, remotePort)), nil
+	}
+
+	return append(ports, remotePort), nil
+}
+
+// waitForExportServiceToBeReady waits until the vmexport service is ready for port-forwarding
+func waitForExportServiceToBeReady(client kubecli.KubevirtClient, vmeInfo *VMExportInfo, interval, timeout time.Duration) (*k8sv1.Service, error) {
+	service := &k8sv1.Service{}
+	serviceName := fmt.Sprintf("virt-export-%s", vmeInfo.Name)
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		vmexport, err := getVirtualMachineExport(client, vmeInfo)
+		if err != nil || vmexport == nil {
+			return false, err
+		}
+
+		if vmexport.Status == nil || vmexport.Status.Phase != exportv1.Ready {
+			fmt.Printf("waiting for VM Export %s status to be ready...\n", vmeInfo.Name)
+			return false, nil
+		}
+
+		service, err = client.CoreV1().Services(vmeInfo.Namespace).Get(context.TODO(), serviceName, metav1.GetOptions{})
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				fmt.Printf("waiting for service %s to be ready before port-forwarding...\n", serviceName)
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	})
+	return service, err
+}
+
+// runPortForward is the actual function that runs the port-forward. Meant to be run concurrently
+func runPortForward(client kubecli.KubevirtClient, pod k8sv1.Pod, namespace string, ports []string, stopChan, readyChan chan struct{}) error {
+	// Create a port forwarding request
+	req := client.CoreV1().RESTClient().Post().
+		Resource("pods").
```

Comment:
Why not point directly to the service?

Reply:
So I initially thought that same thing, but after several failed attempts it turned out to be way more complicated. As you can see [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go#L329), kubectl uses the service to get a pod and calls the actual port-forwarding on it.